Converting project to use Go programming conventions

54v3 opened the pull request:
The original project did not follow typical Go programming practices as outlined in the Go documentation (see https://go.dev/doc/effective_go)

This update converts the project to use Go conventions while keeping the Standard concepts intact.

54v3 committed:
Converting project to use Go programming conventions

The original project did not follow typical Go programming practices as outlined in the Go documentation (see https://go.dev/doc/effective_go)

This update converts the project to use Go conventions while keeping the Standard concepts intact.

hassanhabib commented on core/core.go:
```diff
@@ -0,0 +1,14 @@
+	AddStudent(student Student) Student
+}
+
+type Student struct {
```
What happens when you need to use this model in different services and brokers?
How do you separate brokers from services from models without considering them in a different packages? Is there an equivalent to the concept of namespaces in Go?

54v3 replied:
Any other package can import the core package and use the public types defined within it.  This is what is happening in the brokers and services package already (they use core.Student).

For all intents and purposes a package is a namespace in Go.  Different files within the same package (directory) can interact as if they were in the same file.  See how the mock broker is defined in student_service_mock.go, but used in student_service_test.go as an example.

hassanhabib replied:
@54v3 is there anyway we could have nested packages?
something like: `core.models.Student` and `core.services.StudentService`?

54v3 replied:
You could nest the directories but that doesn't really change anything from the language's perspective.  The name of the package is what is defined at the top of the file, and must be the name of the directory that the file is placed in. core.models.student would be "package student", core.services.StudentService would be package "StudentService".

Conventionally both of those packages would probably want to be one package instead of two, since it is providing functionality related to students.  A teachers package may be a different package, for example.  There is a lot of freedom in how you define packages.  Typically you would envision how someone else might want to use your code in their own project when defining a package.  

ie: Would someone want to use only StudentService, or would a Student always need to be included if I'm using StudentService?  StudentService returns Student, so they should probably be in the same package.  

In this project's example, only the interface is bundled with the Student model while the actual implementation is another package.  This is a typical way to break up packages.  If we wanted to have different implementations of the brokers package instead of calling it "brokers" we could have called it "mysql-brokers" and "redis-brokers", both of which would reference the core package to access the Student model.

hassanhabib replied:
@54v3 Thanks for your feedback.
There are situations where you want to push the same model to different services.
For instance, assume you have `LibraryCardService` and you want to go and say `libraryCardService.CreateLibraryCardForStudent(student)`

This isolation in this case is very important - because you want to be able to reuse your services in separation of the models and in separation of the brokers and so on.

54v3 replied:
Agreed.  I have added LibraryCard functionality to demonstrate how this isolation looks with this project structure.

I included my implementations of LibraryCardService and LibraryCardBroker into the same packages as the StudentService and StudentBroker, but they could be split out into their own packages if required.  Typically you would be grouping the implementations together (for example, you have a MySQL DB and want to implement all of the brokers to store to that backend).

hassanhabib replied:
@54v3 we need to do a YouTube session so we can talk about these issues. I appreciate your time.

hassanhabib commented on services/student_service_test.go:
```diff
@@ -0,0 +1,39 @@
+
+	studentService := NewStudentService(broker)
+
+	actualStudent := studentService.AddStudent(inputStudent)
```
How do you know the type of `actualStudent` by just reading the code?

54v3 replied:
In this case it could be inferred based on seeing the definition of inputStudent, but otherwise you wouldn't be able to determine the type just by looking at it if only shorthand syntax was used.  Usually you can determine the types through naming conventions.  The short hand is optional, you can always use the var x type syntax if you want to.

hassanhabib commented on services/student_service_test.go:
```diff
@@ -0,0 +1,39 @@
+			actualStudent)
+	}
+
+	if broker.callCount != 1 {
```
Is there something similar to Assertions libraries or packages in Go?

54v3 replied:
There are third party libraries that provide some of that functionality.  https://github.com/go-playground/assert might be what you are looking for.  This contains a list of useful libraries for Go:  https://github.com/avelino/awesome-go#testing

Go has built in fuzzing support to help write tests:  https://go.dev/security/fuzz/

hassanhabib replied:
@54v3 Perfect. thank you.

54v3 committed:
Adding LibaryCard functionality

Added core, service, and broker entities to support creation of LibraryCards for a Student
